File: browser-extension/tests/sizeBudget.test.js
// The size + comment ratchet for the extension: no new oversized module, no listed file
// growing, no directory getting comment-heavier. The numbers live in sizeBudget.json and
// only ever go DOWN — a refactor that splits a file lowers its entry (or drops it).
// Scope: .js under browser-extension/src and browser-extension/tests. Paths are repo-relative so every
// surface's budget reads the same.

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { existsSync, readdirSync, readFileSync } from 'node:fs';
import { dirname, join, relative } from 'node:path';
import { fileURLToPath } from 'node:url';
import { spawnSync } from 'node:child_process';

const repoRoot = (() => {
  let dir = dirname(fileURLToPath(import.meta.url));
  while (!existsSync(join(dir, 'core', 'CMakeLists.txt'))) {
    const up = dirname(dir);
    assert.notEqual(up, dir, 'the repo root is above this test');
    dir = up;
  }
  return dir;
})();

const budget = JSON.parse(readFileSync(new URL('./sizeBudget.json', import.meta.url), 'utf8'));
const SCOPE = ['browser-extension/src', 'browser-extension/tests'];
const SKIP_DIRS = new Set(['node_modules', 'build', 'dist', 'third_party']);

const walk = (dir, out = []) => {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    if (entry.name.startsWith('.')) continue;
    const path = join(dir, entry.name);
    if (entry.isDirectory()) { if (!SKIP_DIRS.has(entry.name)) walk(path, out); }
    // .d.ts too: a shape file is source, and the cap has to reach it.
    else if (entry.name.endsWith('.js') || entry.name.endsWith('.d.ts')) out.push(path);
  }
  return out;
};

// The char scan tracks strings so a `//` or `/*` inside a literal doesn't count (` spans lines,
// ' and " don't), and regex literals are consumed whole — a backtick in one would swallow the file.
const REGEX_AFTER = /[(,=:[!&|?{};+\-*%^~<>]\s*$|\b(?:return|typeof|case|in|of|new|do|else|void|delete|instanceof|yield|await)\s*$/;
const endOfRegex = (line, i) => {
  for (let j = i + 1, klass = false; j < line.length; j++) {
    if (line[j] === '\\') { j++; continue; }
    if (klass) { if (line[j] === ']') klass = false; continue; }
    if (line[j] === '[') { klass = true; continue; }
    if (line[j] === '/') return j;
  }
  return -1;
};
const measure = (src) => {
  const lines = src.split('\n');
  if (lines.length && lines[lines.length - 1] === '') lines.pop();
  let comment = 0, inBlock = false, inTemplate = false;
  for (const line of lines) {
    const head = line.trimStart();
    const isComment = inBlock || (!inTemplate && (head.startsWith('//') || head.startsWith('/*')));
    let quote = inTemplate ? '`' : '';
    for (let i = 0; i < line.length; i++) {
      const c = line[i], next = line[i + 1];
      if (inBlock) { if (c === '*' && next === '/') { inBlock = false; i++; } continue; }
      if (quote) {
        if (c === '\\') i++;
        else if (c === quote) quote = '';
        continue;
      }
      if (c === '/' && next === '*') { inBlock = true; i++; continue; }
      if (c === '/' && next === '/') break;
      if (c === '/' && REGEX_AFTER.test(line.slice(0, i))) {
        const end = endOfRegex(line, i);
        if (end > 0) { i = end; continue; }
      }
      if (c === '"' || c === "'" || c === '`') quote = c;
    }
    inTemplate = quote === '`';
    if (isComment) comment++;
  }
  return { total: lines.length, comment };
};

const measured = new Map();
for (const dir of SCOPE) {
  for (const path of walk(join(repoRoot, dir))) {
    measured.set(relative(repoRoot, path), measure(readFileSync(path, 'utf8')));
  }
}
const inScope = [...measured.keys()].sort();
const excepted = (path) => Object.hasOwn(budget.exceptions, path);

test('no unlisted file is over the line cap', () => {
  const over = inScope.filter(
    (p) => !excepted(p) && !(p in budget.files) && measured.get(p).total > budget.maxNewFileLines);
  assert.deepEqual(over, [],
    `over ${budget.maxNewFileLines} lines and unlisted — split it, or record its size in `
    + 'tests/sizeBudget.json (exceptions are for generated bundles and byte-pinned ports only)');
});

test('listed files did not grow', () => {
  const grew = [], shrank = [], gone = [];
  for (const [path, cap] of Object.entries(budget.files)) {
    if (excepted(path)) continue;
    const found = measured.get(path);
    if (!found) { gone.push(path); continue; }
    if (found.total > cap) grew.push(`${path}: ${found.total} > ${cap}`);
    else if (found.total < cap * 0.9) shrank.push(`${path}: ${cap} → ${found.total}`);
  }
  for (const note of shrank) console.log(`  ratchet down: ${note}`);
  for (const note of gone) console.log(`  stale entry (file is gone): ${note}`);
  assert.deepEqual(grew, [], 'listed files may only shrink — split the growth out instead');
});

// A module and its .d.ts are one thing to the reader, so they count once.
const stemsPerDir = () => {
  const dirs = new Map();
  for (const rel of measured.keys()) {
    const at = rel.lastIndexOf('/');
    const dir = rel.slice(0, at);
    const base = rel.slice(at + 1);
    const stem = base.endsWith('.d.ts') ? base.slice(0, -5) : base.slice(0, -3);
    if (!dirs.has(dir)) dirs.set(dir, new Set());
    dirs.get(dir).add(stem);
  }
  return dirs;
};

test('folder fan-out: no directory over the cap, no listed directory grew', () => {
  const cap = budget.maxFilesPerDir;
  const appeared = [], grew = [], shrank = [];
  for (const [dir, stems] of stemsPerDir()) {
    const n = stems.size;
    const allowed = budget.dirs[dir];
    if (allowed === undefined) {
      if (n > cap) appeared.push(`${dir} holds ${n} modules (cap ${cap}) — split it by feature`);
      continue;
    }
    if (n > allowed) grew.push(`${dir} grew to ${n} modules (budget ${allowed})`);
    else if (n < allowed) shrank.push(`'${dir}': ${n}`);
  }
  const gone = Object.keys(budget.dirs).filter((d) => !stemsPerDir().has(d));
  if (shrank.length) console.log(`  note: lower these in sizeBudget.json —\n    ${shrank.join('\n    ')}`);
  assert.deepStrictEqual(appeared, [], 'a new folder over the cap must be split, never recorded');
  assert.deepStrictEqual(grew, [], 'listed folders may shrink, never grow');
  assert.deepStrictEqual(gone, [], 'drop these from sizeBudget.json — the folder is gone');
});

test('comment share per directory did not rise', () => {
  const risen = [];
  for (const [dir, cap] of Object.entries(budget.commentPct)) {
    let total = 0, comment = 0;
    for (const path of inScope) {
      if (dirname(path) !== dir) continue;
      total += measured.get(path).total;
      comment += measured.get(path).comment;
    }
    if (!total) continue;
    const pct = Math.floor(comment * 100 / total);
    if (pct > cap) risen.push(`${dir}: ${pct}% > ${cap}%`);
  }
  assert.deepEqual(risen, [], 'comments outgrew the code — trim the prose, keep the code');
});

// The floor sits ~3% under today's count — raise it when the suite grows a lot.
// STENCIL_TEST_COUNT_RUN marks the inner run, so it never nests.
const TEST_FLOOR = 1580;
const INNER_RUN = 'STENCIL_TEST_COUNT_RUN';

test('test-count floor: the suite still discovers and runs its whole tree',
  { skip: process.env[INNER_RUN] ? 'inner count run' : false }, () => {
    // NODE_TEST_CONTEXT is this process's own runner marker; inherited, it mutes the inner run.
    const innerEnv = { ...process.env, [INNER_RUN]: '1' };
    delete innerEnv.NODE_TEST_CONTEXT;
    const inner = spawnSync(process.execPath, ['--test', '--test-reporter=tap'], {
      cwd: fileURLToPath(new URL('../', import.meta.url)), encoding: 'utf8',
      maxBuffer: 256 * 1024 * 1024, env: innerEnv,
    });
    const total = /^# tests (\d+)$/m.exec(inner.stdout || '');
    assert.ok(total, `the inner runner printed no test total (exit ${inner.status})`);
    const count = Number(total[1]);
    assert.ok(count >= TEST_FLOOR,
      `browser-extension suite collapsed to ${count} tests, floor is ${TEST_FLOOR}`);
  });

File: core/tests/sizeBudget.test.cpp
// Size + comment ratchet for core/: no new oversized file, no listed file growing, no directory
// getting more comment-heavy. The budget is tests/sizeBudget.json, read at RUNTIME.
// Core is codec-free, so this carries a minimal fixed-shape JSON reader of its own (CMake's
// string(JSON) needs 3.19 - the project pins 3.16). Paths resolve from the repo root.
// Print the tree's numbers: core/build/stencil_tests -tc="*budget: measured*" --no-skip
#include "doctest.h"

#include <algorithm>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

namespace {
namespace fs = std::filesystem;

// ── minimal JSON reader (this file only; the budget's shape is fixed) ─────────
struct JsonReader {
  std::string s;
  std::size_t i = 0;

  explicit JsonReader(std::string text) : s(std::move(text)) {}
  void ws() { while (i < s.size() && std::isspace(static_cast<unsigned char>(s[i]))) ++i; }
  bool eat(char c) {
    ws();
    if (i >= s.size() || s[i] != c) return false;
    ++i;
    return true;
  }
  std::string str() {
    std::string out;
    if (!eat('"')) return out;
    for (; i < s.size() && s[i] != '"'; ++i) {
      if (s[i] == '\\' && i + 1 < s.size()) ++i;
      out += s[i];
    }
    ++i;
    return out;
  }
  int num() {
    ws();
    const std::size_t start = i;
    while (i < s.size() && (std::isdigit(static_cast<unsigned char>(s[i])) || s[i] == '-')) ++i;
    return i > start ? std::stoi(s.substr(start, i - start)) : 0;
  }
};

struct Budget {
  int maxNewFileLines = 0;
  int maxFilesPerDir = 0;
  std::map<std::string, int> files;
  std::map<std::string, int> dirs;
  std::map<std::string, int> commentPct;
  std::map<std::string, std::string> exceptions;
};

Budget loadBudget(const fs::path& path) {
  std::ifstream in(path);
  REQUIRE_MESSAGE(in.good(), "missing budget file: " << path.string());
  std::stringstream buf;
  buf << in.rdbuf();
  JsonReader r(buf.str());
  Budget b;
  REQUIRE(r.eat('{'));
  while (!r.eat('}')) {
    const std::string key = r.str();
    REQUIRE(r.eat(':'));
    if (key == "maxNewFileLines") b.maxNewFileLines = r.num();
    else if (key == "maxFilesPerDir") b.maxFilesPerDir = r.num();
    else if (key == "files" || key == "commentPct" || key == "exceptions" || key == "dirs") {
      REQUIRE(r.eat('{'));
      while (!r.eat('}')) {
        const std::string entry = r.str();
        REQUIRE(r.eat(':'));
        if (key == "exceptions") b.exceptions[entry] = r.str();
        else if (key == "files") b.files[entry] = r.num();
        else if (key == "dirs") b.dirs[entry] = r.num();
        else b.commentPct[entry] = r.num();
        r.eat(',');
      }
    } else {
      r.str();  // "_doc"
    }
    r.eat(',');
  }
  REQUIRE(b.maxNewFileLines > 0);
  return b;
}

// ── the tree ─────────────────────────────────────────────────────────────────
fs::path repoRoot() {
  fs::path p = fs::current_path();
  for (int hops = 0; hops < 16; ++hops) {
    if (fs::exists(p / "CLAUDE.md")) return p;
    if (!p.has_parent_path() || p.parent_path() == p) break;
    p = p.parent_path();
  }
  return {};
}

// Scope: .cpp/.hpp/.h under core/, minus vendored headers and build output.
std::vector<fs::path> scopedFiles(const fs::path& root) {
  std::vector<fs::path> out;
  for (fs::recursive_directory_iterator it(root / "core"), end; it != end; ++it) {
    const std::string name = it->path().filename().string();
    if (it->is_directory()) {
      if (name == "third_party" || name.rfind("build", 0) == 0) it.disable_recursion_pending();
      continue;
    }
    const std::string ext = it->path().extension().string();
    if (ext == ".cpp" || ext == ".hpp" || ext == ".h") out.push_back(it->path());
  }
  std::sort(out.begin(), out.end());
  return out;
}

struct Counts { int total = 0; int comment = 0; int files = 0; };

// A comment line opens with // or /*, or sits inside a block comment. String and char
// literals are skipped so a // or /* inside one never flips the state.
Counts countFile(const fs::path& path) {
  std::ifstream in(path, std::ios::binary);
  std::stringstream buf;
  buf << in.rdbuf();
  const std::string src = buf.str();
  Counts c;
  bool inBlock = false;
  for (std::size_t i = 0; i < src.size();) {
    const std::size_t nl = src.find('\n', i);
    const std::size_t end = (nl == std::string::npos) ? src.size() : nl;
    ++c.total;
    std::size_t j = i;
    while (j < end && std::isspace(static_cast<unsigned char>(src[j]))) ++j;
    if (inBlock || (j + 1 < end && src[j] == '/' && (src[j + 1] == '/' || src[j + 1] == '*')))
      ++c.comment;
    for (std::size_t k = i; k < end; ++k) {
      if (inBlock) {
        if (src[k] == '*' && k + 1 < end && src[k + 1] == '/') { inBlock = false; ++k; }
        continue;
      }
      if (src[k] == '"' || src[k] == '\'') {
        const char quote = src[k];
        for (++k; k < end; ++k) {
          if (src[k] == '\\') { ++k; continue; }
          if (src[k] == quote) break;
        }
        continue;
      }
      if (src[k] == '/' && k + 1 < end) {
        if (src[k + 1] == '/') break;
        if (src[k + 1] == '*') { inBlock = true; ++k; }
      }
    }
    if (nl == std::string::npos) break;
    i = nl + 1;
  }
  return c;
}

std::string rel(const fs::path& root, const fs::path& p) {
  return fs::relative(p, root).generic_string();
}

}  // namespace

TEST_CASE("size budget: no core file exceeds its recorded line count") {
  const fs::path root = repoRoot();
  REQUIRE_MESSAGE(!root.empty(), "could not locate the repo root (no CLAUDE.md above cwd)");
  const Budget b = loadBudget(root / "core/tests/sizeBudget.json");

  for (const fs::path& p : scopedFiles(root)) {
    const std::string name = rel(root, p);
    if (b.exceptions.count(name)) continue;  // generated / byte-pinned twins only
    const int lines = countFile(p).total;
    const auto listed = b.files.find(name);
    if (listed == b.files.end()) {
      CHECK_MESSAGE(lines <= b.maxNewFileLines,
                    name << " is " << lines << " lines (> " << b.maxNewFileLines
                         << "): split it, or record it in core/tests/sizeBudget.json");
      continue;
    }
    CHECK_MESSAGE(lines <= listed->second,
                  name << " grew to " << lines << " lines (budget " << listed->second << ")");
    if (lines * 10 < listed->second * 9)
      MESSAGE(name << " shrank to " << lines << " (budget " << listed->second
                   << ") — ratchet the budget down");
  }
}

TEST_CASE("size budget: comment share and folder fan-out per core directory") {
  const fs::path root = repoRoot();
  REQUIRE(!root.empty());
  const Budget b = loadBudget(root / "core/tests/sizeBudget.json");

  std::map<std::string, Counts> byDir;  // files directly in a dir, not its subdirs
  for (const fs::path& p : scopedFiles(root)) {
    const Counts c = countFile(p);
    Counts& d = byDir[rel(root, p.parent_path())];
    d.total += c.total;
    d.comment += c.comment;
    // A header and its .cpp are one module to the reader, so they count once.
    if (p.extension() != ".cpp" || !fs::exists(fs::path(p).replace_extension(".hpp"))) ++d.files;
  }
  for (const auto& e : byDir) {
    const auto f = b.dirs.find(e.first);
    const int cap = f == b.dirs.end() ? b.maxFilesPerDir : f->second;
    CHECK_MESSAGE(e.second.files <= cap, e.first << " holds " << e.second.files
                  << " (cap " << cap << ") — split it");
  }
  for (const auto& entry : b.commentPct) {
    const auto found = byDir.find(entry.first);
    REQUIRE_MESSAGE(found != byDir.end(), "budget lists an empty dir: " << entry.first);
    const int pct = found->second.comment * 100 / found->second.total;
    CHECK_MESSAGE(pct <= entry.second, entry.first << " is " << pct << "% (budget "
                  << entry.second << "%)");
  }
}

TEST_CASE("size budget: measured tree numbers" * doctest::skip()) {
  const fs::path root = repoRoot();
  REQUIRE(!root.empty());
  for (const fs::path& p : scopedFiles(root)) {
    const Counts c = countFile(p);
    MESSAGE("  \"" << rel(root, p) << "\": " << c.total << ",");
  }
}
